examples/selective2d/2d_train.py

```diff
@@ -0,0 +1,354 @@
+  parser.add_argument('--world_size', type=int, default=int(os.environ["WORLD_SIZE"]))
+  parser.add_argument('--device', type=str, default=f"cuda:{os.environ['LOCAL_RANK']}")
+  parser.add_argument('--master_process', type=str_to_bool, default=bool(os.environ['RANK']==0)) 
+  parser.add_argument('--tp_size', type=int, default=2)
+  parser.add_argument('--pp_size', type=int, default=2)
+
+  parser.add_argument('--debug', dest='debug', action='store_true')
+
+  args = parser.parse_args()
+
+  return args
+
+
+def rank_print(x):
+    _rank = os.getenv('RANK')
+    if _rank == '0':
+        print(x)
+
+def get_rand(args):
+    x = torch.randint(0, args.vocab_size, (args.batch_size, args.block_size), device=args.device)
+    y = torch.randint(0, args.vocab_size, (args.batch_size, args.block_size), device=args.device)
+    return x, y
+
+def tp_attention(model, name, mesh, tp_dim=0, q='q', k='k', v='v', o='c_proj'):
+  layer = model.get_submodule(name)
+  parallelize_module(layer, mesh, { q: ColwiseParallel(),
+                                    k: ColwiseParallel(),
+                                    v: ColwiseParallel(),
+                                    o: RowwiseParallel() },
+                                    tp_mesh_dim=tp_dim)
+
+  return model
+
+def tp_mlp(model, name, mesh, tp_dim=0, mlp='mlp'):
+  layer = model.get_submodule(name)
+  parallelize_module(layer, mesh, { mlp: PairwiseParallel() }, tp_mesh_dim=tp_dim)
+
+  return model
+
+
+def tp(model, n_layer, mesh, offset=0, tp_dim=0):
+  for i in range(n_layer):
+    block = model.get_submodule(f'transformer.h.{i + offset}')
+    parallelize_module(block, mesh, {'attn.q': ColwiseParallel(),
+                                     'attn.k': ColwiseParallel(),
+                                     'attn.v': ColwiseParallel(),
+                                     'attn.c_proj': RowwiseParallel(),
+                                     'mlp': PairwiseParallel()},
+                                     tp_mesh_dim=tp_dim) 
+
+  return model
+
+def pp(model, pp_device_mesh, args):
+  from pippy.IR import annotate_split_points, PipeSplitWrapper 
+  from pippy import split_into_equal_size
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_chunks = args.world_size 
+  pp_groups = pp_device_mesh.get_dim_groups()[0]
+
+  output_chunk_spec = (TensorChunkSpec(0), sum_reducer)
+  stage = compile_stage(model, args.rank, args.world_size, pp_chunks, pp_device_mesh, pp_groups, 
+            example_inputs=[X, Y], 
+            output_chunk_spec=output_chunk_spec)
+  
+  print(f'[Rank{_rank}] {stage.submod.print_readable()}')
+  return model, stage 
+
+def pp_and_tp(model, mesh, args):
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_dim, tp_dim = 0, 1
+  pp_rank, tp_rank = args.local_rank // args.tp_size, args.local_rank % args.tp_size
```

Comment:
What if we are doing this on multiple host?

Reply:
in multiple host case, we should use args.rank instead of args.local_rank. i'll change the line to handle multiple hosts in the next commit